src/tracks/playabletrack/wavetrack/ui/WaveChannelView.cpp

```diff
@@ -1541,24 +1542,27 @@ namespace
 
       // Determine whether we should show individual samples
       // or draw circular points as well
-      outAveragePPS = viewRect.width / (rate * (h1 - h));// pixels per sample
-      outShowSamples = outAveragePPS > 0.5;
+      outAvgPixPerSecond = viewRect.width / (h1 - h);
+      const auto pixelsPerSample = outAvgPixPerSecond / sampleRate;
+      outShowSamples = pixelsPerSample > 0.5;
 
       if(outShowSamples)
          // adjustment so that the last circular point doesn't appear
          // to be hanging off the end
-         return  pixelsOffset / (outAveragePPS * rate); // pixels / ( pixels / second ) = seconds
+         return pixelsOffset * stretchRatio /
+                outAvgPixPerSecond; // pixels / ( pixels / second ) = seconds
       return .0;
    }
 }
 
-ClipParameters::ClipParameters
-   (bool spectrum, const SampleTrack *track, const WaveClip *clip, const wxRect &rect,
-   const SelectedRegion &selectedRegion, const ZoomInfo &zoomInfo)
+ClipParameters::ClipParameters(
+   bool spectrum, const SampleTrack* track, const WaveClip* clip,
+   const wxRect& rect, const SelectedRegion& selectedRegion,
+   const ZoomInfo& zoomInfo)
+    : tOffset { clip->GetPlayStartTime() }
+    , sampleRate { static_cast<double>(clip->GetRate()) }
+    , stretchRatio { clip->GetStretchRatio() }
```

Comment:
I am skipping review of the pencil tool for now to work on the commit for spectrum view.

I think this commit and the one for waveform view should be combined into one -- the commit for waveform that I reviewed is not really complete without some of these changes.

As I anticipated, you needed to change `ClipParameters` to remember the sample rate and the stretch, from which a correct sample rate can be computed.

However -- were all the necessary changes made in all places where the old code used rate?

I doubt it, and will give details.